keps/sig-scheduling/3838-pod-mutable-scheduling-directives/README.md

```diff
@@ -0,0 +1,962 @@
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+I want to build a controller that implements workload queueing and influences when and where a 
+workload should run. The workload controller creates the pods of a workload instance. To control 
+when and where the workload should run, I have a webhook that injects a scheduling readiness gate, 
+and a controller that removes the gate when capacity becomes available.
+
+At the time of creating the workload, it may not be known on which subset of the nodes (e.g., a zone or 
+a VM type) a workload should run. To control where a workload should run, the controller updates 
+the node selector/affinity of the workload's pods; by doing so, the queue controller is able to influence the 
+scheduler and autoscaler decisions.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- New calls from a queue controller to the API server to update node affinity/selector. 
+  The mitigation is for such controllers to make a single API call for both updating
+  affinity/selector and removing the scheduling gate.
+
+- App controllers may find it surprising that the pod they created is different from their
+  template. Typically apps don't continusly reconile existing pods
+  to match the template, they react only to change to the pod template in the app's
+  spec. Moreover, the side affects of this proposal is no different from a mutating 
+  webhook that injects affinity onto the pod.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The pod update validation logic in the API server already allows updating labels and 
+annotations, it also allows adding new tolerations. What we need to do is to relax the update validation
+on node affinity/selector.
+
+There are two main problems we need to address:
+
+1) Race conditions between updating node affinity/selector constraints and the kube-scheduler when scheduling
+the pod. As mentioned before, the solution is to restrict the updates to pods with scheduling gates. This
```

Comment:
```suggestion
the pod. As mentioned before, the solution is to restrict the updates to unscheduled pods with scheduling gates. This
```

Reply:
this is more a question for the scheduling gate feature, but do we always validate that spec.nodeName remains unset if the pod still has scheduling gates set (in pod create, update, and bind validation)? it looks like create only does that validation if the feature is enabled, which doesn't look right... that validation should be driven off the populated fields, not the feature gate

Reply:
>> it looks like create only does that validation if the feature is enabled, which doesn't look right... that validation should be driven off the populated fields, not the feature gate
>
> We can adjust that, but the scheduler also ignores gated pods only when the feature is on, should we make it on all the time then?

validation should never permit nodeName and schedulingGates to be populated simultaneously, regardless of feature enablement. That ensures persisted API data is always coherent for downstream components to make use of.

Reply:
looks like BindingREST#setPodHostAndAnnotations and ValidatePodCreate both need to be updated to do their validation unconditionally

Reply:
I added a comment to the issue tracking pod scheduling readiness: https://github.com/kubernetes/kubernetes/issues/113608#issuecomment-1419389539

Reply:
>  That ensures persisted API data is always coherent for downstream components to make use of.

Thanks for raising this. It looks like it will impact vanilla components as well, in an upgrade->rollback scenario like this:

- spin up a 1.26 cluster, but with `PodSchedulingReadiness=true`
- create a pod carrying one scheduling gate
- the pod stays in `SchedulingGated` state
- re-start API Server and scheduler, but with `PodSchedulingReadiness=false`
- (with current logic) the pod enters `Running` state (suppose it requests little resource) while also having non-empty `.spec.schedulingGates` present

So @liggitt @ahg-g, do you think this scenario is concerning? If not, I will pursue the "unconditional validation" change in 1.27; otherwise, we may backport it to 1.26.

Reply:
I would make it in 1.27 and backport